Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs

```diff
@@ -0,0 +1,86 @@
+﻿using UnityEngine;
+using UnityEngine.InputSystem;
+using UnityEngine.InputSystem.Controls;
+using UnityEngine.InputSystem.Layouts;
+using UnityEngine.Scripting;
+using TrackingState = UnityEngine.XR.InputTrackingState;
+
+namespace UnityEngine.InputSystem.XR
+{
+    public struct Pose
+    {
+        public bool isTracked { get; set; }
+        public TrackingState trackingState { get; set; }
+
+        public Vector3 position { get; set; }
+        public Quaternion rotation { get; set; }
+
+        public Vector3 velocity { get; set; }
+        public Vector3 angularVelocity { get; set; }
+    }
```

Comment:
If possible, would recommend making this a `PoseState` struct for `PoseControl` and locking the memory layout of the control down to this struct the same way `TouchControl` does. The reason is that as is, `PoseControl.ReadUnprocessedValueFromState` is *extremely* costly. ATM every `ReadValue` involves several virtual method calls and walking all over the place. In addition, you have several controls that are in turn made up of subcontrols. Reading a single `Vector3`, for example, will involve a `ReadValue` overhead for the vector itself and then a `ReadValue` overhead for three subcontrols.

Some of the `ReadValue` overhead will eventually disappear but even then, I would strongly recommend locking the memory formats of bigger controls down.

Admittedly, that does lead to inconsistencies. For controls that do this, you can do stuff like putting processors on individual child controls but then, when you read a value through the parent control, the process on the child control won't take effect. Though, with how the `PoseControl.ReadUnprocessedValueFromState` looks like ATM you get the same effect (of ignoring processors on child controls).

In the DOTS version, we'll likely have none of these issues as all the data will probably be mass-converted once but for better or worse, in the current implementation, `ReadValue` is not cheap :(

Reply:
Swapped to use a full state block.